scripts/imgtool.py

```diff
@@ -0,0 +1,109 @@
+import argparse
+from imgtool import keys
+from imgtool import image
+
+def gen_rsa2048(args):
+    keys.RSA2048.generate().export_private(args.keyfile)
+def gen_ecdsa_p256(args):
+    keys.ecdsa256p1.generate().export_private(args.keyfile)
+def gen_ecdsa_p224(args):
+    print("TODO: p-224 not yet implemented")
+
+keygens = {
+        'rsa-2048': gen_rsa2048,
+        'ecdsa-p256': gen_ecdsa_p256,
+        'ecdsa-p224': gen_ecdsa_p224, }
+
+def do_keygen(args):
+    if args.type not in keygens:
+        print('Unexpected key type:', args.type)
+        return
+    keygens[args.type](args)
+
+def do_getpub(args):
+    key = keys.load(args.key)
+    key.emit_c()
+
+def do_sign(args):
+    align = args.align
+    # print('sign:', args)
+    img = image.Image.load(args.infile, version=args.version,
+            header_size=args.header_size,
+            pad=args.pad)
+    key = keys.load(args.key)
+    img.sign(key)
+
+    if args.pad:
+        img.pad_to(args.pad, args.align)
+
+    img.save(args.outfile)
+
+subcmds = {
+        'keygen': do_keygen,
+        'getpub': do_getpub,
+        'sign': do_sign, }
+
+def alignment_value(text):
+    value = int(text)
+    if value not in [1, 2, 4, 8]:
+        msg = "{} must be one of 1, 2, 4 or 8".format(value)
+        raise argparse.ArgumentTypeError(msg)
+    return value
+
+def hexint(text):
+    return int(text, 0)
+
+SemiSemVersion = namedtuple('SemiSemVersion', ['major', 'minor', 'revision', 'build'])
+
+version_re = re.compile(r"""^([1-9]\d*)(\.([1-9]\d*)(\.([1-9]\d*)(\+([a-z1-9-][a-z0-9-]*))?)?)?$""")
+def decode_version(text):
+    """Decode the version string, which should be of the form maj.min.rev+build"""
+    # print("decode:", text)
+    m = version_re.match(text)
+    if m:
+        result = SemiSemVersion(
+                int(m.group(1)) if m.group(1) else 0,
+                int(m.group(3)) if m.group(3) else 0,
+                int(m.group(5)) if m.group(5) else 0,
+                int(m.group(7)) if m.group(7) else 0)
+        return result
+    else:
+        msg = "Invalid version number, should be maj.min.rev+build with later parts optional"
+        raise argparse.ArgumentTypeError(msg)
+
+def args():
+    parser = argparse.ArgumentParser()
+    subs = parser.add_subparsers(help='subcommand help', dest='subcmd')
+
+    keygenp = subs.add_parser('keygen', help='Generate pub/private keypair')
+    keygenp.add_argument('-k', '--key', metavar='filename', required=True)
+    keygenp.add_argument('-t', '--type', metavar='type',
+            choices=['rsa-2048', 'ecdsa-p224', 'ecdsa-p256'],
+            required=True)
+
+    getpub = subs.add_parser('getpub', help='Get public key from keypair')
+    getpub.add_argument('-k', '--key', metavar='filename', required=True)
+
+    sign = subs.add_parser('sign', help='Sign an image with a private key')
+    sign.add_argument('-k', '--key', metavar='filename', required=True)
+    sign.add_argument("--align", type=alignment_value, required=True)
```

Comment:
It would be super awesome if this script kept track of these values (alignment, image size, etc.) for all Zephyr targets supported by mcuboot, and accepted a `-b ZEPHYR_BOARD_NAME` instead. That would mean callers wouldn't have to keep track of this stuff. You could still fall back on --align et al if the board was supported out of tree.

Doing this would also let future versions of this script dump this information in machine-readable formats. That could be used to build tooling on top of upstream mcuboot itself, which application build systems could use to build and link Zephyr apps which want to be chain-loaded by mcuboot.

I can help doing this; I've got some of the needed infrastructure already.